### PR TITLE
feat(nx monorepo): initial sdk package

### PR DESCRIPTION
**Pull request by MarkPieszak**
### Description

Initial monorepo - this time with `Nx`

## Testing & creating the Bundle locally

```bash
nvm use
npm i
npm run build:sdk

# (this will run `nx build sdk` (which will emit the type declarations), and then run webpack to build the umd bundle)
```

This will build the package in `dist/packages/sdk` where you'll see it actually creates all of the files & map files. 
image

This is if you just want to see the bundle and everything in action.

---

##  Publishing

From here, we can publish this lib (or other libs once we have them) via a new Nx publish command.

This command requires that we pass a "version", and it automatically auto increments the dist/packages/sdk/package.json file that's created with this version.

```
nx publish sdk --ver=X.X.X

# this will actually run the above scripts 
```

If you look in `packages/sdk/project.json` you'll see the "publish" command in there, which runs the "build" command, and "build:bundle" which just tells Nx to run a custom "script" we have in our sdk package.json (this one runs webpack and builds our UMD bundles).

```
# project.json (in packages/sdk/)

   "publish": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs sdk {args.ver} {args.tag}"
      },
      "dependsOn": [
        {
          "projects": "self",
          "target": "build" 👈 type defs
        },
        {
          "projects": "self",
          "target": "build:bundle" 👈 umd bundle
        }
      ]
    },
    ```
---

### Modifying publish script
    
The script that `nx publish` runs can be found in `/tools/scripts/publish.mjs`. 

### Testing publish locally (without publishing)

If you want to just *test this out* without actually publishing, open up that `publish.mjs` file and comment out the last lines

```
// Execute "npm publish" to publish
// execSync(`npm publish --access public --tag ${tag}`); 👈👈 comment this out to test the bundle getting ready for publish without actually "publishing"
```

Now when you run `nx publish sdk --ver 3.0.0` you can see it builds and gets everything ready, and the package.json within dist/packages/sdk/package.json has that updated 3.0.0 version.

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
image

For building UMD, ESM, CJS, etc we should use `@nrwl/web:rollup` https://nx.dev/packages/web/executors/rollup instead of custom scripts, otherwise Nx won't work as expected as we bring more packages to the monorepo. Rollup should produce both outputs (ESM & UMD) in parallel (when running a single command). Eventually, we should be able to remove all webpack configuration files (currently located in packages/sdk).

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-v16.3.0
+v16.14.0
```
Comment (iamjaredwalters) [inline .nvmrc:1]: Is this version significant somehow?

File: packages/sdk/.env [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# ensure these are added in the key/values in the rollup.config.js file
+API_URL=https://api.us.flatfile.io
+MOUNT_URL=https://app.flatfile.io
```
Comment (iamjaredwalters) [inline packages/sdk/.env:3]: Is `.env` intentionally checked into repo?

File: package.json [hunk 3]
```diff
@@ -1,39 +1,12 @@
...
-  },
-  "homepage": "https://github.com/FlatFilers/sdk#readme",
-  "files": [
-    "dist"
-  ],
-  "main": "dist/index.js",
-  "types": "dist/index.d.ts",
+  "name": "sdk",
+  "version": "0.0.0",
+  "license": "MIT",
```
[the remark this answers is not in the document]
Comment (MarkPieszak) [reply package.json:4]: yes so the "base" package.json doesn't hold any version information, the specific "packages/{whatever}/package.json" has it now.

when we run a publish it asks what version we want to bump it to etc, we don't even have to do it manually anymore - which is great!

File: .nvmrc (hunk 1 above)
Comment (MarkPieszak) [reply .nvmrc:1]: nothing really I just figured to get us on a more recent version!
we can ignore this if need be :)

File: packages/sdk/.env (hunk 2 above)
Comment (MarkPieszak) [reply packages/sdk/.env:3]: it was already checked in previous (at the root of project), and they're used in the code for sdk. webpack just replaces them with these values.

we can hide it if needed! I figured they weren't sensitive, but it does feel weird to show a `.env` - totally agree! :D

**Comment by dboskovic** [comment]
closing in favor of @iamjaredwalters work

**Pull request closed without merging**